### PR TITLE
No tracing memory usage of shared column data in MPPTask's memory tracker

### LINKED ISSUE #8128: Memory tracker is not accurate if read pool is enabled.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1. setup a TiDB cluster with 1 TiFlash node
2. Load TPCH 10 data
3. using the following script to run sql
```
for num in {1..concurrency}
do  
        mysql --host host_ip --port port -uroot < /tmp/test.sql &
done

wait;
```
The test.sql is
```
set tidb_enforce_mpp=1;
use tpch10;
select * from lineitem where l_shipmode like '%ab%';
select * from lineitem where l_shipmode like '%ab%';
select * from lineitem where l_shipmode like '%ab%';
select * from lineitem where l_shipmode like '%ab%';
select * from lineitem where l_shipmode like '%ab%';
```

With read pool disabled, the concurrency is 2, the memory usage reported by memory tracker are
```
[2023/09/21 09:11:23.008 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.04 GiB."] [source=MemoryTracker] [thread_id=243]
[2023/09/21 09:11:23.010 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=242]
[2023/09/21 09:11:23.010 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 2.09 GiB."] [source=MemoryTracker] [thread_id=242]
[2023/09/21 09:11:25.443 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=248]
[2023/09/21 09:11:25.444 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.02 GiB."] [source=MemoryTracker] [thread_id=250]
[2023/09/21 09:11:25.444 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 2.02 GiB."] [source=MemoryTracker] [thread_id=250]
[2023/09/21 09:11:27.851 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=254]
[2023/09/21 09:11:27.892 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.04 GiB."] [source=MemoryTracker] [thread_id=256]
[2023/09/21 09:11:30.229 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.09 GiB."] [source=MemoryTracker] [thread_id=259]
[2023/09/21 09:11:30.265 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.07 GiB."] [source=MemoryTracker] [thread_id=262]
[2023/09/21 09:11:32.590 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.06 GiB."] [source=MemoryTracker] [thread_id=265]
[2023/09/21 09:11:32.635 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.04 GiB."] [source=MemoryTracker] [thread_id=268]
[2023/09/21 09:11:32.635 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 2.04 GiB."] [source=MemoryTracker] [thread_id=268]
```
With read pool disabled, the concurrency is 1, the memory usage reported by memory tracker are
```
[2023/09/21 09:10:46.260 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.06 GiB."] [source=MemoryTracker] [thread_id=223]
[2023/09/21 09:10:46.260 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 1.06 GiB."] [source=MemoryTracker] [thread_id=223]
[2023/09/21 09:10:47.501 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.08 GiB."] [source=MemoryTracker] [thread_id=226]
[2023/09/21 09:10:47.501 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 1.08 GiB."] [source=MemoryTracker] [thread_id=226]
[2023/09/21 09:10:48.741 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.06 GiB."] [source=MemoryTracker] [thread_id=229]
[2023/09/21 09:10:48.742 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 1.06 GiB."] [source=MemoryTracker] [thread_id=229]
[2023/09/21 09:10:50.000 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=232]
[2023/09/21 09:10:50.000 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 1.05 GiB."] [source=MemoryTracker] [thread_id=232]
[2023/09/21 09:10:51.240 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=235]
[2023/09/21 09:10:51.240 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 1.05 GiB."] [source=MemoryTracker] [thread_id=235]

```
With read pool enabled, the concurrency is 2, the memory usage reported by memory tracker are
```
[2023/09/21 09:08:37.087 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 10.48 GiB."] [source=MemoryTracker] [thread_id=269]
[2023/09/21 09:08:37.090 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 288.86 MiB."] [source=MemoryTracker] [thread_id=270]
[2023/09/21 09:08:37.090 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 988.18 MiB."] [source=MemoryTracker] [thread_id=270]
[2023/09/21 09:08:38.407 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 8.28 GiB."] [source=MemoryTracker] [thread_id=275]
[2023/09/21 09:08:38.824 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 244.48 MiB."] [source=MemoryTracker] [thread_id=277]
[2023/09/21 09:08:39.801 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 3.79 GiB."] [source=MemoryTracker] [thread_id=280]
[2023/09/21 09:08:40.381 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 288.98 MiB."] [source=MemoryTracker] [thread_id=283]
[2023/09/21 09:08:41.408 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 2.69 GiB."] [source=MemoryTracker] [thread_id=286]
[2023/09/21 09:08:42.029 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 397.90 MiB."] [source=MemoryTracker] [thread_id=289]
[2023/09/21 09:08:43.009 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 4.41 GiB."] [source=MemoryTracker] [thread_id=292]
[2023/09/21 09:08:43.488 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 435.62 MiB."] [source=MemoryTracker] [thread_id=295]
[2023/09/21 09:08:43.488 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 8.89 GiB."] [source=MemoryTracker] [thread_id=295]
```
With read pool disabled, the concurrency is 1, the memory usage reported by memory tracker are
```
[2023/09/21 09:07:38.645 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 698.27 MiB."] [source=MemoryTracker] [thread_id=252]
[2023/09/21 09:07:38.645 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 698.27 MiB."] [source=MemoryTracker] [thread_id=252]
[2023/09/21 09:07:39.848 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 680.21 MiB."] [source=MemoryTracker] [thread_id=255]
[2023/09/21 09:07:39.848 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 680.21 MiB."] [source=MemoryTracker] [thread_id=255]
[2023/09/21 09:07:41.041 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 685.35 MiB."] [source=MemoryTracker] [thread_id=258]
[2023/09/21 09:07:41.041 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 685.35 MiB."] [source=MemoryTracker] [thread_id=258]
[2023/09/21 09:07:42.242 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 690.88 MiB."] [source=MemoryTracker] [thread_id=261]
[2023/09/21 09:07:42.242 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 690.88 MiB."] [source=MemoryTracker] [thread_id=261]
[2023/09/21 09:07:43.450 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 701.03 MiB."] [source=MemoryTracker] [thread_id=264]
[2023/09/21 09:07:43.450 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 701.03 MiB."] [source=MemoryTracker] [thread_id=264]
```

As we can see, when read pool enabled, when running queries with concurrency(even the concurrent is only 2), the memory usage reported by memory tracker is very inaccurate, for both query level and global level.

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #8128 

### What is changed and how it works?

- In `DMFileReader::readCoumn`, if there are some concurrent `DMFileReader` of the same file existing, this column data is likely to be shared, so disable memory tracing by setting `current_memory_tracker` to `nullptr`.
- Mark `PODArray` the underlying memory of column data to be `is_shared_memory` and use `shared_column_data_mem_tracker` to trace its memory.

Before this PR(from issue #8128)
```
[2023/09/21 09:08:37.087 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 10.48 GiB."] [source=MemoryTracker] [thread_id=269]
[2023/09/21 09:08:37.090 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 288.86 MiB."] [source=MemoryTracker] [thread_id=270]
[2023/09/21 09:08:37.090 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 988.18 MiB."] [source=MemoryTracker] [thread_id=270]
[2023/09/21 09:08:38.407 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 8.28 GiB."] [source=MemoryTracker] [thread_id=275]
[2023/09/21 09:08:38.824 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 244.48 MiB."] [source=MemoryTracker] [thread_id=277]
[2023/09/21 09:08:39.801 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 3.79 GiB."] [source=MemoryTracker] [thread_id=280]
[2023/09/21 09:08:40.381 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 288.98 MiB."] [source=MemoryTracker] [thread_id=283]
[2023/09/21 09:08:41.408 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 2.69 GiB."] [source=MemoryTracker] [thread_id=286]
[2023/09/21 09:08:42.029 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 397.90 MiB."] [source=MemoryTracker] [thread_id=289]
[2023/09/21 09:08:43.009 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 4.41 GiB."] [source=MemoryTracker] [thread_id=292]
[2023/09/21 09:08:43.488 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (for query): 435.62 MiB."] [source=MemoryTracker] [thread_id=295]
[2023/09/21 09:08:43.488 +08:00] [DEBUG] [MemoryTracker.cpp:101] ["Peak memory usage (total): 8.89 GiB."] [source=MemoryTracker] [thread_id=295]
```
After this PR:
```
[2023/09/22 17:20:49.967 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.04 GiB."] [source=MemoryTracker] [thread_id=3084]
[2023/09/22 17:20:53.023 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.60 GiB."] [source=MemoryTracker] [thread_id=3086]
[2023/09/22 17:20:53.132 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.07 GiB."] [source=MemoryTracker] [thread_id=3089]
[2023/09/22 17:20:56.188 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.21 GiB."] [source=MemoryTracker] [thread_id=3091]
[2023/09/22 17:20:56.298 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.02 GiB."] [source=MemoryTracker] [thread_id=3094]
[2023/09/22 17:20:59.324 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.43 GiB."] [source=MemoryTracker] [thread_id=3096]
[2023/09/22 17:20:59.440 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.03 GiB."] [source=MemoryTracker] [thread_id=3098]
[2023/09/22 17:21:02.469 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.28 GiB."] [source=MemoryTracker] [thread_id=3100]
[2023/09/22 17:21:02.564 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.03 GiB."] [source=MemoryTracker] [thread_id=3103]
[2023/09/22 17:21:05.894 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.38 GiB."] [source=MemoryTracker] [thread_id=3105]
[2023/09/22 17:21:05.991 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.01 GiB."] [source=MemoryTracker] [thread_id=3108]
[2023/09/22 17:21:09.070 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.69 GiB."] [source=MemoryTracker] [thread_id=3109]
[2023/09/22 17:21:09.187 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.30 GiB."] [source=MemoryTracker] [thread_id=3112]
[2023/09/22 17:21:12.246 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.25 GiB."] [source=MemoryTracker] [thread_id=3114]
[2023/09/22 17:21:12.361 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.11 GiB."] [source=MemoryTracker] [thread_id=3118]
[2023/09/22 17:21:15.412 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.30 GiB."] [source=MemoryTracker] [thread_id=3120]
[2023/09/22 17:21:15.569 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.04 GiB."] [source=MemoryTracker] [thread_id=3122]
[2023/09/22 17:21:18.653 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.33 GiB."] [source=MemoryTracker] [thread_id=3125]
[2023/09/22 17:21:18.769 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.71 GiB."] [source=MemoryTracker] [thread_id=3128]
[2023/09/22 17:21:21.845 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.28 GiB."] [source=MemoryTracker] [thread_id=3131]
[2023/09/22 17:21:21.988 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.08 GiB."] [source=MemoryTracker] [thread_id=3134]
[2023/09/22 17:21:25.001 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.35 GiB."] [source=MemoryTracker] [thread_id=3137]
[2023/09/22 17:21:25.135 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=3139]
[2023/09/22 17:21:28.144 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.22 GiB."] [source=MemoryTracker] [thread_id=3142]
[2023/09/22 17:21:28.286 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.12 GiB."] [source=MemoryTracker] [thread_id=3143]
[2023/09/22 17:21:31.336 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.20 GiB."] [source=MemoryTracker] [thread_id=3144]
[2023/09/22 17:21:31.454 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.32 GiB."] [source=MemoryTracker] [thread_id=3145]
[2023/09/22 17:21:34.535 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.28 GiB."] [source=MemoryTracker] [thread_id=3146]
[2023/09/22 17:21:34.650 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1023.65 MiB."] [source=MemoryTracker] [thread_id=3148]
[2023/09/22 17:21:37.665 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.31 GiB."] [source=MemoryTracker] [thread_id=3151]
[2023/09/22 17:21:37.799 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.09 GiB."] [source=MemoryTracker] [thread_id=3153]
[2023/09/22 17:21:40.931 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.28 GiB."] [source=MemoryTracker] [thread_id=3155]
[2023/09/22 17:21:41.051 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.20 GiB."] [source=MemoryTracker] [thread_id=3157]
[2023/09/22 17:21:44.169 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.24 GiB."] [source=MemoryTracker] [thread_id=3159]
[2023/09/22 17:21:44.294 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.10 GiB."] [source=MemoryTracker] [thread_id=3162]
[2023/09/22 17:21:47.329 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.25 GiB."] [source=MemoryTracker] [thread_id=3165]
[2023/09/22 17:21:47.441 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.12 GiB."] [source=MemoryTracker] [thread_id=3167]
[2023/09/22 17:21:50.481 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.53 GiB."] [source=MemoryTracker] [thread_id=3170]
[2023/09/22 17:21:50.657 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 998.26 MiB."] [source=MemoryTracker] [thread_id=3171]
[2023/09/22 17:21:53.689 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.67 GiB."] [source=MemoryTracker] [thread_id=3173]
[2023/09/22 17:21:53.801 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.88 GiB."] [source=MemoryTracker] [thread_id=3176]
[2023/09/22 17:21:56.932 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.36 GiB."] [source=MemoryTracker] [thread_id=3179]
[2023/09/22 17:21:57.026 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.06 GiB."] [source=MemoryTracker] [thread_id=3181]
[2023/09/22 17:22:00.085 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.62 GiB."] [source=MemoryTracker] [thread_id=3183]
[2023/09/22 17:22:00.210 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.13 GiB."] [source=MemoryTracker] [thread_id=3185]
[2023/09/22 17:22:03.338 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.35 GiB."] [source=MemoryTracker] [thread_id=3186]
[2023/09/22 17:22:03.421 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.06 GiB."] [source=MemoryTracker] [thread_id=3188]
[2023/09/22 17:22:06.471 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.48 GiB."] [source=MemoryTracker] [thread_id=3191]
[2023/09/22 17:22:06.589 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (for query): 1.05 GiB."] [source=MemoryTracker] [thread_id=3193]
[2023/09/22 17:22:06.589 +08:00] [DEBUG] [MemoryTracker.cpp:107] ["Peak memory usage (total): 3.86 GiB."] [source=MemoryTracker] [thread_id=3193]
```

- Maybe there are some global PODArray object that are not being traced before, about 12.4MiB.
image

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by JaySon-Huang — COMMENTED** [review]
Rest LGTM

**Review by JaySon-Huang — APPROVED** [review]
LGTM

File: dbms/src/Common/MemoryTracker.cpp [hunk 1]
```diff
@@ -111,7 +117,11 @@ void MemoryTracker::alloc(Int64 size, bool check_memory_limit)
     reportAmount();
 
     if (!next.load(std::memory_order_relaxed))
+    {
         CurrentMetrics::add(metric, size);
+        if (shared_column_data_mem_tracker)
```
Comment (windtalker) [inline dbms/src/Common/MemoryTracker.cpp:122]: Maybe data race here?

File: dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp [hunk 2]
```diff
@@ -809,8 +810,14 @@ void DMFileReader::readColumn(
...
     size_t skip_packs)
 {
+    bool has_concurrent_reader = DMFileReaderPool::instance().hasConcurrentReader(*this);
     if (!getCachedPacks(column_define.id, start_pack_id, pack_count, read_rows, column))
     {
+        // If there are concurrent read requests, this data is likely to be shared.
+        // So the allocation and deallocation of this data may not be in the same MemoryTracker.
+        // This can lead to inaccurate memory statistics of MemoryTracker.
+        // To solve this problem, we use a independent global memory tracker to trace the shared column data in ColumnSharingCacheMap.
+        MemoryTrackerSetter mem_tracker_guard(true, has_concurrent_reader ? nullptr : current_memory_tracker);
```
Comment (windtalker) [inline dbms/src/Storages/DeltaMerge/File/DMFileReader.cpp:820]: Don't have to use this mem_tracker_guard if `has_concurrent_reader` is false?

File: dbms/src/Storages/DeltaMerge/ReadThread/ColumnSharingCache.cpp [hunk 3]
```diff
@@ -61,6 +61,14 @@ void DMFileReaderPool::set(DMFileReader & from_reader, int64_t col_id, size_t st
     }
 }
 
+// Check is there any concurrent DMFileReader with `from_reader`.
+bool DMFileReaderPool::hasConcurrentReader(DMFileReader & from_reader)
+{
+    std::lock_guard lock(mtx);
+    auto itr = readers.find(from_reader.path());
+    return itr != readers.end() && itr->second.size() >= 2;
```
Comment (windtalker) [inline dbms/src/Storages/DeltaMerge/ReadThread/ColumnSharingCache.cpp:69]: So actually, there is a chance that the reader becomes a `ConcurrentReader` after `hasConcurrentReader` is called?

File: dbms/src/Common/MemoryTracker.cpp (hunk 1 above)
Comment (windtalker) [reply dbms/src/Common/MemoryTracker.cpp:122]: Yes, I aggree, but theoretically there will be data race here, not sure if our ASan tests will complain it or not. I'm ok with current implementation, and I think we can fix it if it breaks the ASan tests.

**Review by windtalker — APPROVED** [review]
LGTM

**Pull request merged**